### PR TITLE
Update `has_inputs_and_outputs()` for new consensus rules

### LINKED ISSUE #2365: Check enableSpends or enableOutputs in has_inputs_and_outputs
## Motivation

The NU5 consensus rules were updated to check `enableSpends` or `enableOutputs` when counting orchard inputs or outputs.

### Specifications

image

## Solution

- [ ] update the consensus rules in the `check::has_inputs_and_outputs` comment
- [ ] update the code in `check::has_inputs_and_outputs`
- [ ] add or update tests for this new behaviour

## Related Work

Orchard was added in PR #2229

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We need to update Zebra for new consensus rules added to the protocol in regards to orchard flags. Closes https://github.com/ZcashFoundation/zebra/issues/2365

### Specifications

* At least one of `tx_in_count`, `nSpendsSapling`, and `nJoinSplit` MUST be non-zero.
* At least one of `tx_out_count`, `nOutputsSapling`, and `nJoinSplit` MUST be non-zero.
* This condition must hold: `tx_in_count` > 0 or `nSpendsSapling` > 0 or (`nActionsOrchard` > 0 and `enableSpendsOrchard` = 1)
* This condition must hold: `tx_out_count` > 0 or `nOutputsSapling` > 0 or (`nActionsOrchard` > 0 and `enableOutputsOrchard` = 1)

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

## Solution

Add the new consensus rules into the `has_inputs_and_outputs` function, update the tests to cover them.

## Review

Anyone can review the code initially, before merging @teor2345 should take a look.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-consensus/src/transaction/tests.rs
Comment (oxarbitrage) [inline zebra-consensus/src/transaction/tests.rs:97]: I think there should be no need to get a new transaction with a new call to `insert_fake_orchard_shielded_data()` each time we want to test something however i had troubles with the mutable borrows and passing them multiple times.

Comment (teor2345) [reply zebra-consensus/src/transaction/tests.rs:97]: You should feel free to clone transactions whenever you need to.

Or maybe @jvff can help here?

File: zebra-consensus/src/transaction/check.rs
Comment (teor2345) [inline zebra-consensus/src/transaction/check.rs:48]: Confusing code: The actions lines are a bit long and confusing.

Do you think we should split them out into their own variables?

**Review by teor2345 — DISMISSED** [review]
This PR is good enough to merge now. I'll leave it to @oxarbitrage to make some improvements if he wants.

File: zebra-consensus/src/transaction/tests.rs
Comment (daira) [inline zebra-consensus/src/transaction/tests.rs]:
```suggestion
    // The check will fail if the transaction has no flags
```

File: zebra-consensus/src/transaction/check.rs
Comment (daira) [inline zebra-consensus/src/transaction/check.rs:41]: I don't think there can be an arithmetic overflow in practice, but still, it's not expressing the rule in the same way as the spec.

File: zebra-consensus/src/transaction/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/transaction/tests.rs:97]: I created a low priority ticket for this (https://github.com/ZcashFoundation/zebra/issues/2409) and i added a TODO comment into the code for it in https://github.com/ZcashFoundation/zebra/pull/2398/commits/05019a9d4d5b7a18517489029f9a9a19171b76c4

File: zebra-consensus/src/transaction/check.rs
Comment (oxarbitrage) [reply zebra-consensus/src/transaction/check.rs:48]: I created a ticket (https://github.com/ZcashFoundation/zebra/issues/2410) for this and added a TODO comment into the code in https://github.com/ZcashFoundation/zebra/pull/2398/commits/05019a9d4d5b7a18517489029f9a9a19171b76c4

File: zebra-consensus/src/transaction/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/transaction/tests.rs]: Thanks, i fixed this comment as part of https://github.com/ZcashFoundation/zebra/pull/2398/commits/05019a9d4d5b7a18517489029f9a9a19171b76c4

File: zebra-consensus/src/transaction/check.rs
Comment (oxarbitrage) [reply zebra-consensus/src/transaction/check.rs:41]: Thanks, i created a ticket to improve this code https://github.com/ZcashFoundation/zebra/issues/2410

**Comment by oxarbitrage** [comment]
I moved the remaining items to new lower priority tickets and enabled auto merge here. Need a new approval.

**Review by teor2345 — APPROVED** [review]
This has tests and a follow up ticket, so it's good to go

**Pull request merged**
